core/miner/storage_fsm/impl/precommit_batcher_impl.cpp

```diff
@@ -0,0 +1,156 @@
+      const std::shared_ptr<libp2p::protocol::Scheduler>& scheduler,
+      const Address &miner_address) {
+    std::shared_ptr<PreCommitBatcherImpl> batcher =
+        std::make_shared<PreCommitBatcherImpl>(
+            max_wait, std::move(api), miner_address, max_wait);
+
+    batcher->cutoff_start_ = std::chrono::system_clock::now();
+    batcher->logger_ = common::createLogger("batcher");
+    batcher->logger_->info("Bather have been started");
+    batcher->handle_ = scheduler->schedule(max_wait, [=]() {
+      auto maybe_result = batcher->sendBatch();
+      for (const auto &[key, cb] : batcher->callbacks_) {
+        cb(maybe_result);
+      }
+      batcher->callbacks_.clear();
+      batcher->handle_.reschedule(max_wait);
+    });
+    return batcher;
+  }
+
+  outcome::result<CID> PreCommitBatcherImpl::sendBatch() {
+    std::unique_lock<std::mutex> locker(mutex_, std::defer_lock);
+    // TODO(Elestrias): [FIL-398] goodFunds = mutualDeposit + MaxFee; -  for
+    // checking payable
+    CID message_cid;
+    if (batch_storage_.size() != 0) {
+      logger_->info("Sending procedure started");
+
+      OUTCOME_TRY(head, api_->ChainHead());
+      OUTCOME_TRY(minfo, api_->StateMinerInfo(miner_address_, head->key));
+
+      PreCommitBatch::Params params = {};
+
+      for (const auto &data : batch_storage_) {
+        mutual_deposit_ += data.second.deposit;
+        params.sectors.push_back(data.second.precommit_info);
+      }
+
+      OUTCOME_TRY(encodedParams, codec::cbor::encode(params));
+
+      OUTCOME_TRY(signed_message,
+                  api_->MpoolPushMessage(
+                      vm::message::UnsignedMessage(
+                          miner_address_,
+                          minfo.worker,  // TODO: handle worker
+                          0,
+                          mutual_deposit_,
+                          {},
+                          {},
+                          25,  // TODO (m.tagirov) Miner actor v5 PreCommitBatch
+                          MethodParams{encodedParams}),
+                      kPushNoSpec));
+
+      mutual_deposit_ = 0;
+      batch_storage_.clear();
+      message_cid = signed_message.getCid();
+      logger_->info("Sending procedure completed");
+      cutoff_start_ = std::chrono::system_clock::now();
+      return message_cid;
+    }
+    cutoff_start_ = std::chrono::system_clock::now();
+    return ERROR_TEXT("Empty Batcher");
+  }
+
+  void PreCommitBatcherImpl::forceSend() {
+    auto maybe_result = sendBatch();
+    for (const auto &[key, cb] : callbacks_) {
+      cb(maybe_result);
+    }
+    callbacks_.clear();
+    handle_.reschedule(max_delay_);
+  }
+
+  void PreCommitBatcherImpl::setPreCommitCutoff(const ChainEpoch &current_epoch,
+                                                const SectorInfo &sector_info) {
+    ChainEpoch cutoff_epoch =
+        sector_info.ticket_epoch
+        + static_cast<int64_t>(fc::kEpochsInDay + kChainFinality);
```

Comment:
This code is already in fc namespace, why do you need fc:: ?

Reply:
agree